curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-regular-expressions-by-building-a-spam-filter/66bb6467b4b7381178420970.md

```diff
@@ -0,0 +1,182 @@
+---
+id: 66bb6467b4b7381178420970
+title: Step 20
+challengeType: 0
+dashedName: step-20
+---
+
+# --description--
+
+While this expression does match `1 hundred dollars`, it will not however match `1  hundred  dollars`. Spam messages
+can and will find a way to bypass flaws in our detector. Let's add a preventive measure and replace the hard-coded
+single spaces with a `\s+` expression. The `\s` represents a universal space character and the `+` means one 
+or more of the previous characters. 
```

Comment:
I'm confused. Can you explain why it won't match `10 dollars` after this step? The intended solution is for dollarRegex to be equal to `/[0-9]+s\+(?:hundred|thousand|million|billion)?\s+dollars/i` after this step. 

As for the intent, the point is to keep things mentally clear. A plus sign before an literal empty space can be confusing and obnoxious to deal with. While distracted, it's really easy to get confused and ask "Hey what's that space doing?". Then hit backspace and get stuck on the challenge. Plus when trying to come up with an explanation of how the plus sign works, I'd like there to be a proper symbol to be there. I wanna be able to point at what's being repeated.

Reply:
Oh yeah, that makes sense for space. It might be good to add more a bit more which characters other than literal space `\s` includes.

> I'm confused. Can you explain why it won't match `10 dollars` after this step? The intended solution is for dollarRegex to be equal to `/[0-9]+\s+(?:hundred|thousand|million|billion)?\s+dollars/i` after this step.

There's `\s+` before and after the optional group. That's one or more spaces twice.

Reply:
Okay. Now that I can see the issue, what's the best way to resolve the problem?

Reply:
Not exactly sure, it's easy to make it much more complicated quickly: 
```regex
/[0-9]+\s+(?:(hundred|thousand|million|billion)\s+)?dollars/i
```

What do you think about changing the first `\s+` to `\s*`? 
```regex
/[0-9]+\s*(?:hundred|thousand|million|billion)?\s+dollars/i
```
Matching ie. `2hundred dollars` wouldn't be that bad, would it?

Reply:
I suppose it wouldn't be so bad.